### DISCUSSION TITLE
Container migration

### DISCUSSION THREAD

**Discussion by hathibelagal-dev** [InterDigitalInc/AdvantEDGE#227] [Q&A] [answered]
Can the AdvantEDGE platform be used to test the performance of different container migration strategies? For example, for a given scenario, can I compare how much quicker `docker export` is compared to `docker save`?

**Comment by roymx** [comment, marked as answer]
Hello @hathibelagal-dev,

Depends what the use case you want to experiment is.

To purely compare performance of `docker export` vs `docker save` on a specific Docker deployment, you do not need AdvantEDGE.

However, if you implement a container migration framework where components run on different edge nodes and communicate together, and you want to test/measure the overall framework performance in different network conditions/topologies -- then I believe AdvantEDGE can help.

One point of interest, version 1.7 should be out at the end of April/first week of May -  new API endpoints will be added to allow controlling pod lifecylce (e.g terminating/starting pods) at scenario runtime.
This could facilitate the experimentation of a container migration framework.

Thank you for your question and I recon that my answer is generic, so don't hesitate to be more specific on the pursued objectives/scenario if you can - so I can issue more precise guidance.

**Reply by hathibelagal-dev** [reply]
Hello @roymx 

I'm currently trying to test different container migration strategies for V2X services. Simply put, I'm moving containers from one edge host to another as vehicles move along a path, and measuring the downtimes. According to your wiki, I'm following deployment model 1, which means I have one edge application for one UE.

I'm trying to understand if AdvantEDGE can:

- allow me to specify how a container is moved from one edge host to another
- tell me that an edge host change is imminent

Thank you.

**Reply by roymx** [reply]
Hello @hathibelagal-dev ,

Cool use case!

- [ ] **With regards to telling AdvantEDGE "how a container is moved"** -- latest released version of the platform (v1.6.1) does not allow to create new containers at runtime. So use a scenario with pre-instantiated edge applications (source & target) and perform the context transfer between these applications.
As mentioned in my previous message, version 1.7 (end of April) will have a new API that allows instantiating containers at runtime. Of course, your software will have to trigger the REST API to get the container instantiated or alternatively you can use the frontend to create/delete containers. 

- [ ] **With regards to knowing that an edge host change is imminent** - this is tied to predictability and is not _readily available_ from the platform, but there is still some information that can support predictability.
In the model, under `PhysicalLocation` there is a field called `networkLocationsInRange` -- This field will contain the PoAs in range of a UE. In order to get `networkLocationsInRange` populated, you need to enable the `PoAs In Range` automation (see events/automations toolbar in the frontend) - Active scenario can be obtained using the `GET /active` of the Sandbox Controller pod.
Using PoAs in range with the UE `PhysicalLocation/geoData` should allow to develop a simple predictability algorithm.
A more involved approach would consist in modifying GIS Engine to add a new endpoint - this would require to be familiar with GIS concepts & GIS database query language but it would simplify geospatial calculations.
Finally (& not helping for now), the Location Service has an endpoint called `/queries/distance` that would provide the distance between a UE and another point such a PoA - this is not supported now but should be by the summer. Unclear if we can include it in the April release at this point. 

Hope it helps.

**Reply by hathibelagal-dev** [reply]
That definitely helps. Thank you.
